src/katgpucbf/xbgpu/engine.py

```diff
@@ -254,25 +266,180 @@ async def sender_loop(self) -> None:
+                affinity=engine.dst_affinity,
+                comp_vector=engine.dst_comp_vector,
+                stream_config=stream_config,
+                buffers=buffers,
+            ),
+            tx_enabled=init_tx_enabled,
+        )
+
+        self._populate_sensors()
+
+    def _populate_sensors(self) -> None:
+        sensors = self.engine.sensors
+        # Dynamic sensors
+        for output in self.outputs:
+            sensors.add(
+                aiokatcp.Sensor(
+                    int,
+                    f"{output.name}-beng-clip-cnt",
+                    "Number of complex samples that saturated.",
+                    default=0,
+                    initial_status=aiokatcp.Sensor.Status.NOMINAL,
+                )
+            )
+
+    async def gpu_proc_loop(self) -> None:  # noqa: D102
+        while True:
+            # Get item from the receiver loop.
+            # - Wait for the HtoD transfers to complete, then
+            # - Give the chunk back to the receiver for reuse.
+            rx_item = await self._rx_item_queue.get()
+            if rx_item is None:
+                break
+            await rx_item.async_wait_for_events()
+
+            tx_item = await self._tx_free_item_queue.get()
+            await tx_item.async_wait_for_events()
+            tx_item.reset(rx_item.timestamp)
+
+            # Bind input buffers
+
+            # Queue GPU work
+            # - For now, just fill it with zeros
+            tx_item.buffer_device.zero(self._proc_command_queue)
+            tx_item.saturated.zero(self._proc_command_queue)
+
+            # Bind output buffers
+            tx_item.add_marker(self._proc_command_queue)
+            self._tx_item_queue.put_nowait(tx_item)
+
+            # Finish with the rx_item
+            rx_item.add_marker(self._proc_command_queue)
+            self.engine.free_rx_item(rx_item)
+        # When the stream is closed, if the sender loop is waiting for a tx item,
+        # it will never exit. Upon receiving this NoneType, the sender_loop can
+        # stop waiting and exit.
+        logger.debug("gpu_proc_loop completed")
+        self._tx_item_queue.put_nowait(None)
+
+    async def sender_loop(self) -> None:  # noqa: D102
+        # NOTE: This function passes the entire downloaded data to
+        # chunk.send, which then takes care of directing data to each beam's
+        # output destination.
+
+        while True:
+            # Get a populated TxQueueItem from the tx_item_queue
+            item = await self._tx_item_queue.get()
+            if item is None:
+                break
+            # The CPU doesn't need to wait, but the GPU does to ensure it
+            # won't start the download before computation is done.
+            item.enqueue_wait_for_events(self._download_command_queue)
+
+            # Get a free Chunk from the send_stream
+            chunk = await self.send_stream.get_free_chunk()
+
+            # Kick off any data downloads from the GPU
```

Comment:
```suggestion
            # Kick off data downloads from the GPU
```
"Any" sounds like there is doubt about whether there be some or not.